sstables/sstable_directory.cc

```diff
@@ -398,6 +431,19 @@ future<> sstable_directory::sstables_registry_components_lister::process(sstable
     });
 }
 
+future<> sstable_directory::restore_components_lister::process(sstable_directory& directory, process_flags flags) {
+    co_await coroutine::parallel_for_each(_toc_filenames, [flags, &directory] (sstring toc_filename) -> future<> {
+        std::filesystem::path sst_path{toc_filename};
+        entry_descriptor desc = sstables::parse_path(sst_path, "", "");
+        if (!sstable_generation_generator::maybe_owned_by_this_shard(desc.generation)) {
+            co_return;
+        }
+        desc.prefix = sst_path.parent_path().native();
```

Comment:
It looks like desc.prefix is used to carry it through process_descriptor() up (or down) to load_sstable(). The former is one of just two places where the latter is called, the other place is load_foreign_sstables() , but in case of restore it's not called. 

Next, the only difference that load_sstable() makes for desc with and without prefix cases is which storage options it uses. By default (no prefix on desc) it uses this->_storage_options, otherwise it cooks special purpose options.

So instead of putting the prefix on entry_descritpr, how about passing the std::function (or noncopyable_function) into process_descriptor() and later on to load_sstable() that would make storage options to call `_manager.make_sstable()` with? Existing places would pass there `[this] { return this->_storage_options; }` and the restore_components_lister::process() will pass there more sophisticated lambda that will cook correct options.